### PR TITLE
[SPARK-25769][SQL]make UnresolvedAttribute.sql escape nested columns correctly

### PR DESCRIPTION
**Pull request by huaxingao**
## What changes were proposed in this pull request?
Currently, the nested columns are not escaped correctly:
```
scala> $"a.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res0: String = `a.b`

scala> $"`a.b`".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res1: String = `a.b`   // ambiguous

scala> $"`a`.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res2: String = `a.b`    // ambiquous

scala> $"`a.b`.c".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res3: String = ```a.b``.c`    // two verbose
```
It should be something as the following:
```
scala> $"a.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res0: String = a.b

scala> $"`a.b`".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res1: String = `a.b`

scala> $"`a`.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res2: String = a.b

scala> $"`a.b`.c".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res3: String = `a.b`.c
```

## How was this patch tested?
Add test

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
@huaxingao . I like this improvement because of the following four examples.

**BEFORE**
```scala
scala> $"a.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res0: String = `a.b`

scala> $"`a.b`".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res1: String = `a.b`   // ambiguous

scala> $"`a`.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res2: String = `a.b`    // ambiquous

scala> $"`a.b`.c".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res3: String = ```a.b``.c`    // two verbose
```

**AFTER**
```scala
scala> $"a.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res0: String = `a`.`b`

scala> $"`a.b`".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res1: String = `a.b`

scala> $"`a`.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res2: String = `a`.`b`

scala> $"`a.b`.c".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res3: String = `a.b`.`c`
```

Since the updated test result doesn't clealy show your contribution, could you add some test cases covering the above four examples explicitly? It will prevent future regression, too.

**Comment by viirya** [comment]
@huaxingao Can you update the PR description based on @dongjoon-hyun's comments https://github.com/apache/spark/pull/22788#pullrequestreview-166797320. It is more clear on the improvement of this change.

**Comment by viirya** [comment]
@huaxingao Can you update the PR description based on @dongjoon-hyun's comments https://github.com/apache/spark/pull/22788#pullrequestreview-166797320. It is more clear on the improvement of this change.

File: sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out [hunk 1]
```diff
@@ -81,7 +81,7 @@ SELECT t1.i1 FROM t1, mydb1.t1
 struct<>
 -- !query 9 output
 org.apache.spark.sql.AnalysisException
-Reference 't1.i1' is ambiguous, could be: mydb1.t1.i1, mydb1.t1.i1.; line 1 pos 7
+Reference '`t1`.`i1`' is ambiguous, could be: mydb1.t1.i1, mydb1.t1.i1.; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: why is the new format better? it's more verbose, isn't it?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -2702,7 +2702,7 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
 
       val e = intercept[AnalysisException](sql("SELECT v.i from (SELECT i FROM v)"))
       assert(e.message ==
-        "cannot resolve '`v.i`' given input columns: [__auto_generated_subquery_name.i]")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I think the problem here is the out-most backticks, do you know where we add it?

File: sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: I thought some examples in [the above](https://github.com/apache/spark/pull/22788#pullrequestreview-166797320).

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 2 above)
Comment (huaxingao) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: The out-most backticks is added in 
```case _```, in ```case class UnresolvedAttribute(nameParts: Seq[String])```
```
  override def sql: String = name match {
    case ParserUtils.escapedIdentifier(_) | ParserUtils.qualifiedEscapedIdentifier(_, _) => name
    case _ => quoteIdentifier(name)
  }
```
the ```nameParts``` is a Seq of ```"v" "i"```, name is ```v.i```

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 3]
```diff
@@ -99,7 +99,7 @@ case class UnresolvedTableValuedFunction(
 case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute with Unevaluable {
 
   def name: String =
-    nameParts.map(n => if (n.contains(".")) s"`$n`" else n).mkString(".")
+    nameParts.map(n => if (nameParts.length > 1 || n.contains(".")) s"`$n`" else n).mkString(".")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: I don't think this is better for `name`, we should update `sql` though.

File: sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: These examples only make sense when we have the outer backticks. e.g. `'t1.i1'` is good.

**Comment by dbtsai** [comment]
@cloud-fan @dongjoon-hyun instead of changing `Filter` API, do you think using proper escaped char like this PR in https://github.com/apache/spark/pull/22573 is a good approach?

**Comment by cloud-fan** [comment]
Yea I think so, we can even use JSON to be safer. e.g. for `a.b.c.d`, we can encode it as a json array [a,b,c,d]. At data source side, use a json parser to read it back.

**Comment by dbtsai** [comment]
@cloud-fan I like the idea of using JSON, but that will also change the definition of string format. Do we just use JSON for nested case so the existing data source doesn't have to be changed?

**Comment by cloud-fan** [comment]
yea, only use json if it's a nested column.

**Comment by viirya** [comment]
From above examples,

```scala
scala> $"`a.b`".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
res1: String = `a.b`   // ambiguous
```

Is this ambiguous? Doesn't it mean a column simply named `a.b`? Btw I don't see the difference between before and after this change.

**Comment by dongjoon-hyun** [comment]
@viirya . Please see the all four examples. I guess you missed the context.
- https://github.com/apache/spark/pull/22788#pullrequestreview-166797320

BTW, I'm good for any methods if we can proceed further, @cloud-fan and @dbtsai .

**Comment by cloud-fan** [comment]
I agree with the problem described in the PR description that `UnresolvedAttribute.sql` is not ideal. But we should just update `UnresolvedAttribute.sql`, not the `name` method. `name` is used in other places and I think it has no problem.

**Comment by viirya** [comment]
@dongjoon-hyun Oh I see. The ambiguousness is in the results of `sql` for several inputs.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2856,6 +2856,21 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+
+    val sql2 = $"`a.b`".expr.asInstanceOf[UnresolvedAttribute].sql
+    assert(sql2 === "`a.b`")
+
+    val sql3 = $"`a`.b".expr.asInstanceOf[UnresolvedAttribute].sql
+    assert(sql3 === "`a`.`b`")
+
+    val sql4 = $"`a.b`.c".expr.asInstanceOf[UnresolvedAttribute].sql
+    assert(sql4 === "`a.b`.`c`")
+  }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Hi, @huaxingao . Can we move this test to `ColumnExpressionSuite`? I made a [PR](https://github.com/huaxingao/spark/pull/1) to you. Please review and merge.

File: sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out [hunk 5]
```diff
@@ -161,7 +161,7 @@ SELECT db1.t1.i1 FROM t1, mydb2.t1
 struct<>
 -- !query 18 output
 org.apache.spark.sql.AnalysisException
-cannot resolve '`db1.t1.i1`' given input columns: [mydb2.t1.i1, mydb2.t1.i1]; line 1 pos 7
+cannot resolve '`db1`.`t1`.`i1`' given input columns: [mydb2.t1.i1, mydb2.t1.i1]; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: do you think we should just make `sql` same as `name`? It looks to me that `'db1.t1.i1'` is better than `` '`db1`.`t1`.`i1`' ``, as it's more compact and is not ambiguous.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (huaxingao) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Merged. Thank you very much! @dongjoon-hyun

File: sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out (hunk 5 above)
Comment (huaxingao) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: Yes. I agree. For the four examples, we will have the following results:
```
$"a.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
```
with my previous fix: 
```
`a`.`b` 
```     
make sql same as name: 
```
a.b
```
```
$"`a.b`".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
```
with my previous fix: 
```
`a`.`b`
```      
make sql same as name: 
```
`a.b`
```
```
$"`a`.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
```
with my previous fix: 
```
`a`.`b`     
```
make sql same as name: 
```
a.b
```
```
$"`a.b`.c".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
```
with my previous fix: 
```
`a.b`.`c`
```     
make sql same as name: 
```
`a.b`.c         
```
Does this look good to everybody?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: LGTM

Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: @huaxingao Were you planning to make a change to make `name` same as `sql` ? Currently they are different ?

Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: @cloud-fan I had a question, the 3rd example from huaxin, wanted to confirm again if the output looks okay to you.

**from huaxin**
```
$"`a`.b".expr.asInstanceOf[org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute].sql
```
with my previous fix:
```
`a`.`b`
```     
make sql same as name:
```
a.b
```
Is it okay to drop the backtick from the first identifier ?

Comment (huaxingao) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: @dilipbiswal 
Currently ```sql``` is not the same as ```name```

Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: @cloud-fan Makes sense. Thanks for clarification...

**Comment by huaxingao** [comment]
I have a question regarding the test failure in ```ExpressionTypeCheckingSuite```. Most of the tests in this suite failed after I change ```UnresolvedAttribute.sql = UnresolvedAttribute.name```. For example, 
```
  test("check types for unary arithmetic") {
    assertError(BitwiseNot('stringField), "requires integral type")
  }
```
The test failed with
```
"cannot resolve '~`stringField`' due to data type mismatch: argument 1 requires integral
 type, however, '`stringField`' is of string type.;" did not contain "cannot resolve
 '~stringField' due to data type mismatch:"
```
It seems that the root cause of the failure is that ```UnresolveAttribute.sql``` doesn't match with ```AttributeReference.sql``` any more after my fix. I am in doubt if it is  correct to make ```UnresolvedAttribute.sql ``` the same as ``` UnresolvedAttribute.name```

**Comment by huaxingao** [comment]
@cloud-fan @dongjoon-hyun 
Because of the above test failures in ```ExpressionTypeCheckingSuite```, shall I revert to the previous change ?
```
  override def sql: String = nameParts.map { part =>
    part match {
      case ParserUtils.escapedIdentifier(_) | ParserUtils.qualifiedEscapedIdentifier(_, _) =>
        part
      case _ =>
        quoteIdentifier(part)
    }
  }.mkString(".")
```
Or change ```ExpressionTypeCheckingSuite```?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I'm retriggering this to see the test failures.

**Pull request closed without merging**